File: unity/Rotgrid/Assets/Scripts/Systems/MysteryBox.cs
using System.Collections.Generic;
using UnityEngine;

namespace Rotgrid
{
    /// <summary>
    /// The mystery crate. Spins for a rarity-weighted weapon, then occasionally
    /// packs up and relocates so its position is never a given.
    /// </summary>
    public class MysteryBox : Interactable
    {
        enum State { Idle, Spin, Reveal, Leaving, Hidden }

        const float SpinTime = 2.9f;
        const float GrabTime = 9f;
        const int BaseCost = 950;

        readonly WorldBuilder _world;
        readonly Transform _parent;
        readonly GameObject _root;
        readonly MachineInfo _info;
        readonly List<WeaponDef> _pool;
        readonly GameObject _displayRoot;
        GameObject _displayModel;

        State _state = State.Idle;
        float _t, _cycleT, _fireSaleT;
        int _usesLeft;
        int _spotIndex;
        public WeaponDef pending;

        public MysteryBox(WorldBuilder world, InteractionSystem interaction, Transform parent)
            : base("box", Vector3.zero, 2.6f, 2)
        {
            _world = world;
            _parent = parent;
            _pool = Weapons.BoxPool();
            _usesLeft = Util.RandInt(4, 8);

            _root = Props.MysteryBox(parent);
            _info = _root.GetComponent<MachineInfo>();
            _displayRoot = new GameObject("display");
            _displayRoot.transform.SetParent(_root.transform, false);
            _displayRoot.transform.localPosition = new Vector3(0, 1.55f, 0);
            _displayRoot.SetActive(false);

            interaction.Add(this);
            MoveTo(0, true, null);
        }

        public int Cost { get { return _fireSaleT > 0f ? 10 : BaseCost; } }
        BoxSpotDef Spot { get { return MapData.BoxSpots[_spotIndex]; } }

        public void StartFireSale(float seconds) { _fireSaleT = seconds; }

        void MoveTo(int index, bool immediate, GameRun run)
        {
            _spotIndex = index;
            var s = Spot;
            _root.transform.position = new Vector3(s.x, s.y, s.z);
            _root.transform.rotation = Quaternion.Euler(0, s.rot, 0);
            _root.transform.localScale = Vector3.one;
            _root.SetActive(true);
            pos = new Vector3(s.x, s.y + 0.6f, s.z);
            _usesLeft = Util.RandInt(4, 8);
            _state = State.Idle;
            if (!immediate)
            {
                Game.Audio.PlayAt(Sfx.BoxStop, new Vector3(s.x, s.y + 1f, s.z), 1f);
                if (run != null) run.Toast("THE CRATE HAS MOVED", ToastKind.Warn, false);
            }
        }

        void Relocate(GameRun run)
        {
            var options = new List<int>();
            for (int i = 0; i < MapData.BoxSpots.Count; i++)
                if (i != _spotIndex && _world.IsZoneActive(MapData.BoxSpots[i].zone)) options.Add(i);
            MoveTo(options.Count > 0 ? options[Random.Range(0, options.Count)] : _spotIndex, false, run);
        }

        public override Prompt GetPrompt(GameRun run)
        {
            if (_state == State.Spin) return Prompt.Blocked("CRATE CYCLING", "");
            if (_state == State.Reveal) return Prompt.Plain("TAKE " + pending.name, pending.RarityLabel);
            if (_state != State.Idle) return Prompt.Blocked("CRATE SEALED", "");
            return Prompt.Buy("MYSTERY CRATE", Cost,
                _fireSaleT > 0f ? "FIRE SALE" : _usesLeft + " CYCLES LEFT");
        }

        public override bool Activate(GameRun run)
        {
            if (_state == State.Reveal)
            {
                run.weapons.GiveWeapon(pending);
                pending = null;
                _displayRoot.SetActive(false);
                if (_usesLeft <= 0) { _state = State.Leaving; _t = 2f; }
                else { _state = State.Idle; _t = 0f; }
                return true;
            }
            if (_state != State.Idle) return false;
            if (!run.player.Spend(Cost)) { run.Deny(); return false; }
            run.player.stats.boxSpins++;
            _state = State.Spin;
            _t = SpinTime;
            _cycleT = 0f;
            _usesLeft--;
            _displayRoot.SetActive(true);
            Game.Audio.PlayAt(Sfx.BoxSpin, _root.transform.position, 1f);
            return true;
        }

        WeaponDef Roll(int round)
        {
            // Higher rounds tilt the pool toward better hardware.
            float bias = Mathf.Clamp01((round - 4) / 22f);
            var weights = new float[_pool.Count];
            for (int i = 0; i < _pool.Count; i++)
            {
                int order = (int)_pool[i].rarity;
                weights[i] = _pool[i].boxWeight * (1f + order * bias * 1.6f) * (order == 4 ? 0.55f + bias : 1f);
            }
            return _pool[Util.WeightedPick(weights)];
        }

        void SetDisplay(WeaponDef def)
        {
            if (_displayModel != null) Object.Destroy(_displayModel);
            _displayModel = Props.WeaponModel(def, 1.15f, _displayRoot.transform);
        }

        public void Update(float dt, float time, int round, GameRun run)
        {
            if (_fireSaleT > 0f) _fireSaleT -= dt;
            var lid = _info.pivot;
            var light = _info.light;

            switch (_state)
            {
                case State.Spin:
                    _t -= dt;
                    _cycleT -= dt;
                    if (lid != null) lid.localRotation = Quaternion.Euler(-Mathf.Clamp01((SpinTime - _t) * 3f) * 72f, 0, 0);
                    if (light != null) light.intensity = 48f + Mathf.Sin(time * 22f) * 22f;
                    if (_cycleT <= 0f)
                    {
                        // Slow the carousel down as it settles.
                        float p = 1f - _t / SpinTime;
                        _cycleT = 0.07f + p * p * 0.36f;
                        SetDisplay(Util.Pick(_pool));
                    }
                    SpinDisplay(time, 3.2f, 6f);
                    if (_t <= 0f)
                    {
                        pending = Roll(round);
                        SetDisplay(pending);
                        _state = State.Reveal;
                        _t = GrabTime;
                        bool rare = (int)pending.rarity >= 3;
                        Game.Audio.PlayAt(rare ? Sfx.BoxStopRare : Sfx.BoxStop, _root.transform.position, 1f);
                        if (run != null) run.Toast(pending.name, rare ? ToastKind.Warn : ToastKind.Good, false);
                    }
                    break;

                case State.Reveal:
                    _t -= dt;
                    if (light != null) light.intensity = 40f + Mathf.Sin(time * 6f) * 12f;
                    SpinDisplay(time, 1.6f, 3f);
                    if (_t <= 0f)
                    {
                        _displayRoot.SetActive(false);
                        pending = null;
                        if (_usesLeft <= 0) { _state = State.Leaving; _t = 2f; }
                        else { _state = State.Idle; _t = 0f; }
                        if (lid != null) lid.localRotation = Quaternion.identity;
                    }
                    break;

                case State.Leaving:
                {
                    _t -= dt;
                    float k = Mathf.Clamp01(1f - _t / 2f);
                    var s = Spot;
                    _root.transform.position = new Vector3(s.x, s.y + k * k * 9f, s.z);
                    _root.transform.Rotate(0, dt * 290f, 0);
                    _root.transform.localScale = Vector3.one * (1f - k * 0.5f);
                    if (_t <= 0f) { _root.SetActive(false); _state = State.Hidden; _t = 1.2f; }
                    break;
                }

                case State.Hidden:
                    _t -= dt;
                    if (_t <= 0f) Relocate(run);
                    break;

                default:
                    if (lid != null) lid.localRotation = Quaternion.Lerp(lid.localRotation, Quaternion.identity, Mathf.Clamp01(dt * 6f));
                    if (light != null)
                        light.intensity = _fireSaleT > 0f
                            ? 34f + Mathf.Sin(time * 12f) * 16f
                            : 9f + Mathf.Sin(time * 1.4f) * 4f;
                    break;
            }
        }

        void SpinDisplay(float time, float spin, float bobRate)
        {
            if (_displayModel == null) return;
            _displayModel.transform.localRotation = Quaternion.Euler(0, time * spin * Mathf.Rad2Deg * 0.1f, 0);
            _displayModel.transform.localPosition = new Vector3(0, Mathf.Sin(time * bobRate) * 0.05f, 0);
        }

        public void Reset()
        {
            _state = State.Idle;
            _fireSaleT = 0f;
            _displayRoot.SetActive(false);
            MoveTo(0, true, null);
        }
    }
}

File: src/systems/MysteryBox.js
import * as THREE from 'three';
import { makeMysteryBox, makeWeaponModel } from '../world/Props.js';
import { Interactable } from './Interaction.js';
import { MAP } from '../world/MapData.js';
import { boxPool, RARITY } from '../weapons/WeaponDefs.js';
import { Audio } from '../core/AudioEngine.js';
import { rand, randInt, pick, weightedPick, clamp } from '../core/Util.js';

const SPIN_TIME = 2.9;
const GRAB_TIME = 9.0;
const BASE_COST = 950;

const STATE = { IDLE: 0, SPIN: 1, REVEAL: 2, LEAVING: 3, HIDDEN: 4 };

/**
 * The mystery crate. Spins for a random weapon, occasionally packs up and
 * relocates to another anchor so its position is never a given.
 */
export class MysteryBox {
  constructor(world, scene, interaction, ctxRef) {
    this.world = world;
    this.scene = scene;
    this.ctxRef = ctxRef;
    this.pool = boxPool();
    this.state = STATE.IDLE;
    this.t = 0;
    this.usesLeft = randInt(4, 8);
    this.fireSaleT = 0;
    this.pendingWeapon = null;

    this.group = makeMysteryBox();
    scene.add(this.group);

    this.display = new THREE.Group();
    this.display.visible = false;
    this.group.add(this.display);
    this.displayModel = null;

    this.beam = new THREE.Mesh(
      new THREE.CylinderGeometry(0.5, 0.12, 6, 12, 1, true),
      new THREE.MeshBasicMaterial({
        color: 0x9fd93a, transparent: true, opacity: 0, side: THREE.DoubleSide,
        blending: THREE.AdditiveBlending, depthWrite: false,
      }),
    );
    this.beam.position.y = 3.2;
    this.group.add(this.beam);

    const self = this;
    this.interactable = interaction.add(new (class extends Interactable {
      prompt(ctx) {
        if (self.state === STATE.SPIN) return { text: 'CRATE CYCLING', blocked: true };
        if (self.state === STATE.REVEAL) {
          return { text: `TAKE ${self.pendingWeapon.name}`, sub: RARITY[self.pendingWeapon.rarity].label };
        }
        if (self.state !== STATE.IDLE) return { text: 'CRATE SEALED', blocked: true };
        return { text: 'MYSTERY CRATE', cost: self.cost, sub: self.fireSaleT > 0 ? 'FIRE SALE' : `${self.usesLeft} CYCLES LEFT` };
      }
      activate(ctx) { return self.activate(ctx); }
    })({ id: 'box', x: 0, y: 0, z: 0, radius: 2.6, priority: 2 }));

    this.spotIndex = 0;
    this.moveTo(MAP.boxSpots[0], true);
  }

  get cost() { return this.fireSaleT > 0 ? 10 : BASE_COST; }
  get spot() { return MAP.boxSpots[this.spotIndex]; }

  /** Anchors inside zones the player can currently reach. */
  availableSpots() {
    return MAP.boxSpots.filter((s) => this.world.isZoneActive(s.zone));
  }

  moveTo(spot, immediate = false) {
    this.spotIndex = MAP.boxSpots.indexOf(spot);
    this.group.position.set(spot.x, spot.y, spot.z);
    this.group.rotation.y = spot.rot;
    this.interactable.pos.set(spot.x, spot.y + 0.6, spot.z);
    this.usesLeft = randInt(4, 8);
    this.state = STATE.IDLE;
    this.group.visible = true;
    this.group.scale.setScalar(1);
    if (!immediate) {
      Audio.playAt('boxstop', spot, { volume: 1 });
      this.ctxRef()?.toast('THE CRATE HAS MOVED', 'warn');
    }
  }

  relocate() {
    const options = this.availableSpots().filter((s) => s !== this.spot);
    const next = options.length ? pick(options) : this.spot;
    this.moveTo(next);
  }

  startFireSale(seconds) {
    this.fireSaleT = seconds;
  }

  activate(ctx) {
    if (this.state === STATE.REVEAL) {
      ctx.giveWeapon(this.pendingWeapon);
      this.state = STATE.IDLE;
      this.display.visible = false;
      this.pendingWeapon = null;
      this.t = 0;
      if (this.usesLeft <= 0) {
        this.state = STATE.LEAVING;
        this.t = 2.0;
      }
      return true;
    }
    if (this.state !== STATE.IDLE) return false;
    if (!ctx.player.spend(this.cost)) { ctx.deny(); return false; }
    ctx.player.stats.boxSpins++;
    this.state = STATE.SPIN;
    this.t = SPIN_TIME;
    this.usesLeft--;
    this.cycleT = 0;
    this.display.visible = true;
    Audio.playAt('box', this.group.position, { volume: 1 });
    return true;
  }

  _rollWeapon(round) {
    // Higher rounds tilt the pool toward better hardware.
    const bias = clamp((round - 4) / 22, 0, 1);
    const entries = this.pool.map((w) => {
      const order = RARITY[w.rarity].order;
      const weight = w.boxWeight * (1 + order * bias * 1.6) * (order === 4 ? 0.55 + bias : 1);
      return { weight, w };
    });
    return weightedPick(entries).w;
  }

  _setDisplay(def) {
    if (this.displayModel) this.display.remove(this.displayModel);
    this.displayModel = makeWeaponModel(def, 1.15);
    this.displayModel.position.set(0, 1.55, 0);
    this.display.add(this.displayModel);
  }

  update(dt, time, round) {
    if (this.fireSaleT > 0) this.fireSaleT -= dt;
    const lid = this.group.userData.lid;
    const glow = this.group.userData.glow;

    switch (this.state) {
      case STATE.SPIN: {
        this.t -= dt;
        this.cycleT -= dt;
        lid.rotation.x = -clamp((SPIN_TIME - this.t) * 3, 0, 1) * 1.25;
        glow.intensity = 48 + Math.sin(time * 22) * 22;
        this.beam.material.opacity = 0.16 + Math.sin(time * 18) * 0.06;
        if (this.cycleT <= 0) {
          // Slow the carousel down as it settles.
          const p = 1 - this.t / SPIN_TIME;
          this.cycleT = 0.07 + p * p * 0.36;
          this._setDisplay(pick(this.pool));
        }
        if (this.displayModel) {
          this.displayModel.rotation.y = time * 3.2;
          this.displayModel.position.y = 1.55 + Math.sin(time * 6) * 0.05;
        }
        if (this.t <= 0) {
          this.pendingWeapon = this._rollWeapon(round);
          this._setDisplay(this.pendingWeapon);
          this.state = STATE.REVEAL;
          this.t = GRAB_TIME;
          const rare = RARITY[this.pendingWeapon.rarity].order >= 3;
          Audio.playAt('boxstop', this.group.position, { volume: 1, rare });
          this.ctxRef()?.toast(this.pendingWeapon.name, rare ? 'warn' : 'pts');
        }
        break;
      }
      case STATE.REVEAL: {
        this.t -= dt;
        glow.intensity = 40 + Math.sin(time * 6) * 12;
        this.beam.material.opacity = 0.12;
        if (this.displayModel) {
          this.displayModel.rotation.y = time * 1.6;
          this.displayModel.position.y = 1.55 + Math.sin(time * 3) * 0.06;
        }
        if (this.t <= 0) {
          this.display.visible = false;
          this.pendingWeapon = null;
          this.state = this.usesLeft <= 0 ? STATE.LEAVING : STATE.IDLE;
          this.t = this.usesLeft <= 0 ? 2.0 : 0;
          lid.rotation.x = 0;
        }
        break;
      }
      case STATE.LEAVING: {
        this.t -= dt;
        const k = clamp(1 - this.t / 2.0, 0, 1);
        this.group.position.y = this.spot.y + k * k * 9;
        this.group.rotation.y += dt * 5;
        this.group.scale.setScalar(1 - k * 0.5);
        this.beam.material.opacity = 0.3 * (1 - k);
        if (this.t <= 0) {
          this.group.visible = false;
          this.state = STATE.HIDDEN;
          this.t = 1.2;
        }
        break;
      }
      case STATE.HIDDEN: {
        this.t -= dt;
        if (this.t <= 0) this.relocate();
        break;
      }
      default: {
        lid.rotation.x += (0 - lid.rotation.x) * Math.min(1, dt * 6);
        glow.intensity = this.fireSaleT > 0 ? 34 + Math.sin(time * 12) * 16 : 9 + Math.sin(time * 1.4) * 4;
        this.beam.material.opacity = this.fireSaleT > 0 ? 0.14 : 0.05;
        break;
      }
    }
  }

  reset() {
    this.state = STATE.IDLE;
    this.fireSaleT = 0;
    this.display.visible = false;
    this.moveTo(MAP.boxSpots[0], true);
  }
}

export { rand };
